x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,156 @@
+        this.cacheService = Objects.requireNonNull(cacheService);
+        this.cacheDir = Files.createDirectories(cacheDir);
+    }
+
+    public void close() throws IOException {
+        super.close();
+        // Ideally we could let the cache evict/remove cached files by itself after the
+        // directory has been closed.
+        cacheService.removeFromCache(key -> key.startsWith(cacheDir.toString()));
+    }
+
+    @Override
+    public IndexInput openInput(final String name, final IOContext context) throws IOException {
+        ensureOpen();
+        return new CacheBufferedIndexInput(name, fileLength(name), context);
+    }
+
+    public class CacheBufferedIndexInput extends BufferedIndexInput {
+
+        private final String fileName;
+        private final long fileLength;
+        private final Path file;
+        private final IOContext ioContext;
+        private final long offset;
+        private final long end;
+
+        /** a releasable provided by the CacheService that must be released when this IndexInput is done with the cached file **/
+        private @Nullable Releasable releasable;
+        private List<IndexInput> clones;
+        private boolean closed;
+
+        CacheBufferedIndexInput(String fileName, long fileLength, IOContext ioContext) {
+            this(fileName, fileLength, ioContext, "CachedBufferedIndexInput(" + fileName + ")", 0L, fileLength);
+        }
+
+        private CacheBufferedIndexInput(String fileName, long fileLength, IOContext ioContext, String desc, long offset, long length) {
+            super(desc, ioContext);
+            this.fileName = fileName;
+            this.fileLength = fileLength;
+            this.file = cacheDir.resolve(fileName);
+            this.ioContext = ioContext;
+            this.offset = offset;
+            this.end = offset + length;
+            this.clones = new ArrayList<>();
+        }
+
+        @Override
+        public long length() {
+            return end - offset;
+        }
+
+        @Override
+        protected void readInternal(final byte[] b, final int off, final int len) throws IOException {
+            final long pos = getFilePointer() + offset;
+            releasable =
+                cacheService.readFromCache(file, fileLength, releasable, () -> in.openInput(fileName, ioContext), pos, b, off, len);
+        }
+
+        @Override
+        protected void seekInternal(long pos) throws IOException {
+            if (pos > length()) {
+                throw new EOFException("Reading past end of file [position=" + pos + ", length=" + length() + "] for " + toString());
+            } else if (pos < 0L) {
+                throw new IOException("Seeking to negative position [" + pos + "] for " + toString());
+            }
+        }
+
+        @Override
+        public CacheBufferedIndexInput clone() {
+            final CacheBufferedIndexInput clone = (CacheBufferedIndexInput) super.clone();
+            clone.clones = new ArrayList<>();
```

Comment:
I think we clone and/or slice from different threads so we should make this threadsafe:

```suggestion
            clone.clones = new CopyOnWriteArrayList <>();
```